packages/aws-cdk/lib/cdk-toolkit.ts

```diff
@@ -243,6 +248,82 @@ export class CdkToolkit {
+        'Make sure to add a "watch" key to your cdk.json');
+    }
+
+    // For the "include" subkey under the "watch" key, the behavior is:
+    // 1. No "watch" setting? We error out.
+    // 2. "watch" setting without an "include" key? We default to observing "./**".
+    // 3. "watch" setting with an empty "include" key? We default to observing "./**".
+    // 4. Non-empty "include" key? Just use the "include" key.
+    const watchIncludes = this.patternsArrayForWatch(watchSettings.include, rootDir, /* returnRootDirIfEmpty */ true);
+    debug("'include' patterns for 'watch': %s", watchIncludes);
+
+    // For the "exclude" subkey under the "watch" key,
+    // the behavior is to add some default excludes in addition to the ones specified by the user:
+    // 1. The CDK output directory.
+    // 2. Any file whose name starts with a dot.
+    // 3. Any directory's content whose name starts with a dot.
+    // 4. Any node_modules and its content.
+    const outputDir = this.props.configuration.settings.get(['output']);
+    const watchExcludes = this.patternsArrayForWatch(watchSettings.exclude, rootDir, /* returnRootDirIfEmpty */ false).concat(
+      `${rootDir}/${outputDir}/**`,
+      `${rootDir}/**/.*`,
+      `${rootDir}/**/.*/**`,
+      `${rootDir}/**/node_modules/**`,
+    );
+    debug("'exclude' patterns for 'watch': %s", watchExcludes);
+
+    // Since 'cdk deploy' is a relatively slow operation for a 'watch' process,
+    // introduce a concurrency latch that tracks the state.
+    // This way, if file change events arrive when a 'cdk deploy' is still executing,
+    // we will batch them, and trigger another 'cdk deploy' after the current one finishes,
+    // making sure 'cdk deploy's  always execute one at a time.
+    let latch: 'pre-ready' | 'open' | 'deploying' | 'queued' = 'pre-ready';
+    chokidar.watch(watchIncludes, {
+      ignored: watchExcludes,
+      cwd: rootDir,
+      // ignoreInitial: true,
+    }).on('ready', () => {
+      latch = 'open';
+      debug("'watch' received the 'ready' event. From now on, all file changes will trigger a deployment");
+    }).on('all', async (event, filePath) => {
+      if (latch === 'pre-ready') {
+        print(`'watch' is observing ${event === 'addDir' ? 'directory' : 'the file'} '%s' for changes`, filePath);
+      } else if (latch === 'open') {
+        latch = 'deploying';
+        print("Detected change to '%s' (type: %s). Triggering 'cdk deploy'", filePath, event);
+        await this.invokeDeployFromWatch(options);
+
+        // If latch is still 'deploying' after the 'await', that's fine,
+        // but if it's 'queued', that means we need to deploy again
+
+        // TypeScript doesn't realize latch can change between 'awaits',
+        // and thinks the below 'while' condition is always false -
+        // so, suppress the compiler for that line
+        // @ts-ignore
+        while (latch === 'queued') {
```

Comment:
The `@ts-ignore` can be avoided with:
```ts
while (latch as 'deploying' | 'queued' === 'queued') { // typecast to express that it can be `deploying` or `queued` after the `await`
  // ...
}
```

or with
```ts
let latch: 'pre-ready' | 'open' | 'deploying' | 'queued' = 'pre-ready';
const getLatch = () => latch;
// ...
// ...
while (getLatch() === 'queued') {
  // ...
}
```

Reply:
Thanks @jogold. I don't like the second one, but I'll use the first one instead of of turning off the compiler completely.